packages/contractkit/src/identity/offchain/utils.ts

```diff
@@ -0,0 +1,350 @@
+ * @param toAddress address to encrypt symmetric key to
+ */
+const distributeSymmetricKey = async (
+  wrapper: OffchainDataWrapper,
+  dataPath: string,
+  key: Buffer,
+  toAddress: Address
+): Promise<void | OffchainErrors> => {
+  const accounts = await wrapper.kit.contracts.getAccounts()
+  const [fromPubKey, toPubKey] = await Promise.all([
+    accounts.getDataEncryptionKey(wrapper.self),
+    accounts.getDataEncryptionKey(toAddress),
+  ])
+  const wallet = wrapper.kit.getWallet()
+  const sharedSecret = await wallet.computeSharedSecret(publicKeyToAddress(fromPubKey), toPubKey)
+
+  const computedDataPath = getCiphertextLabel(`${dataPath}.key`, sharedSecret, fromPubKey, toPubKey)
+  const encryptedData = Encrypt(
+    Buffer.from(trimUncompressedPrefix(ensureUncompressed(toPubKey)), 'hex'),
+    key
+  )
+
+  const signature = await signBuffer(wrapper, computedDataPath, encryptedData)
+  return wrapper.writeDataTo(
+    encryptedData,
+    Buffer.from(trimLeading0x(signature), 'hex'),
+    computedDataPath
+  )
+}
+
+/**
+ * Handles choosing the symmetric key to use.
+ * If we're explicitly passing in a key, use that,
+ * If a key has already been generated for this dataPath, use that,
+ * Else generate a new one
+ *
+ * @param wrapper the offchain data wrapper
+ * @param dataPath path to where the encrypted data is stored. Used to derive the key location
+ * @param symmetricKey
+ */
+async function fetchOrGenerateKey(
+  wrapper: OffchainDataWrapper,
+  dataPath: string,
+  symmetricKey?: Buffer
+) {
+  if (symmetricKey) {
+    return Ok(symmetricKey)
+  }
+
+  const existingKey = await readSymmetricKey(wrapper, dataPath, wrapper.self)
+  if (existingKey.ok) {
+    return Ok(existingKey.result)
+  }
+
+  if (
+    existingKey.error.errorType === SchemaErrorTypes.OffchainError &&
+    existingKey.error.error.errorType === OffchainErrorTypes.NoStorageRootProvidedData
+  ) {
+    return Ok(randomBytes(16))
+  }
+
+  return Err(existingKey.error)
+}
+
+/**
+ * Handles encrypting the data with a symmetric key, then distributing said key to each address
+ * in the `toAddresses` array.
+ *
+ * @param wrapper the offchain data wrapper
+ * @param dataPath path to where the encrypted data is stored. Used to derive the key location
+ * @param data the data to encrypt
+ * @param toAddresses the addresses to distribute the symmetric key to
+ * @param symmetricKey the symmetric key to use to encrypt the data. One will be found or generated if not provided
+ */
+export const writeEncrypted = async (
+  wrapper: OffchainDataWrapper,
+  dataPath: string,
+  data: Buffer,
+  toAddresses: Address[],
+  symmetricKey?: Buffer
+): Promise<SchemaErrors | void> => {
+  const fetchKey = await fetchOrGenerateKey(wrapper, dataPath, symmetricKey)
+  if (!fetchKey.ok) {
+    return fetchKey.error
+  }
+
+  const iv = randomBytes(16)
+  const payload = AES128Encrypt(fetchKey.result, iv, data)
+  const signature = await signBuffer(wrapper, `${dataPath}.enc`, payload)
```

Comment:
I bet we already handle it, though not obvious to me directly - do we care about people "stealing" data? This ciphertext is signed but an attacker can take the ciphertext and replace the signature with their own. 

Specifically I can also set my data encryption key to be w/e I want, if I'm reading correctly: https://github.com/celo-org/celo-monorepo/blob/54197498c3890674f7068aca3b8b39d5d709a3f4/packages/protocol/contracts/common/Accounts.sol#L165

I didn't dive deeper into it since it's probably already thought about by someone, so I'd be happy to understand where it fails.

Reply:
Yes the storage provider could be malicious and replace data, that is true of anywhere your data is stored that you don't own. In my mind this is alleviated by the fact that when you're requesting Bob's name, you always verify the signature is from his account's public key or an authorized signer, let me know if I'm missing something.

Yes that's also true your DEK can be any buffer >= 33 bytes. The doc says `secp256k1 public key for data encryption. Preferably compressed` which @nategraf has mentioned before. It's true that we could add additional checks to this function but that's out of scope for this PR.

Reply:
I'm imagining the following:
* I'm another user who wants to "steal" data and present it as my own.
* I take the encrypted blob and replace the signature with my account key.
* I set my data encryption key to be the person I stole it from.

Reply:
@kobigurk are you talking about the situation where a malicious user has write access to storage? I've spent a decent amount of time thinking about this and I think we're still OK. The DEK is not used for signing, the attacker cannot simply replace the signature with theirs. It needs to be done with an authorized signer or account key.

Otherwise there's nothing stopping people from claiming things that aren't true. I believe the case you're thinking of is covered by verifying on-chain.

##### Name claim
- User A claims their name
- X "steals" their name
- B tries to read A's name and see's A
- when they read X's name they also see A. Not the end of the world

##### TX comment
 - User A sends a TX to B with a private off-chain comment 
- X "steals" their TX comment
- B receives a transaction, reads the comment from (`<User A storage root>/<txHash>/comment`) and sees the correct TX comment
- if A decided to read the TX comment from X's storage root they also see the comment data

Do those handle what you're thinking of?

Reply:
I think they do! Thanks for walking me through it.